frontend/pages/shopping-lists/_id.vue

```diff
@@ -525,16 +539,31 @@ export default defineComponent({
         return labelSetting;
       });
 
-      // setting this doesn't have any effect on the data since it's refreshed automatically, but it makes the ux feel smoother
-      shoppingList.value.labelSettings = labelSettings;
-      updateListItemOrder();
+      localLabels.value = labelSettings
+    }
+
+    function cancelLabelOrder() {
+      loadingCounter.value -= 1
+      if (!shoppingList.value) {
+        return;
+      }
+      // restore original state
+      localLabels.value = shoppingList.value.labelSettings
+    }
+
+    async function saveLabelOrder() {
+      if (!shoppingList.value || !localLabels.value || (localLabels.value === shoppingList.value.labelSettings)) {
+        return;
+      }
 
       loadingCounter.value += 1;
-      const { data } = await userApi.shopping.lists.updateLabelSettings(shoppingList.value.id, labelSettings);
+      const { data } = await userApi.shopping.lists.updateLabelSettings(shoppingList.value.id, localLabels.value);
       loadingCounter.value -= 1;
 
       if (data) {
-        refresh();
+        // update shoppingList labels using the API response
+        shoppingList.value.labelSettings = (data as ShoppingListOut).labelSettings
```

Comment:
Missing a semicolon here

```suggestion
        shoppingList.value.labelSettings = (data as ShoppingListOut).labelSettings;
```